File: lab-redes/java/tcp/ServidorTCP.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class ServidorTCP {
    private static final int PORTA_BASE = 5000;
    private static final DateTimeFormatter FORMATO_HORA = DateTimeFormatter.ofPattern("HH:mm:ss");

    public static void main(String[] args) throws IOException {
        int porta = PORTA_BASE + lerOffset(args);

        try (ServerSocket servidor = new ServerSocket(porta)) {
            System.out.println("[TCP] Servidor aguardando conexoes na porta " + porta + "...");

            try (Socket cliente = servidor.accept();
                 BufferedReader entrada = new BufferedReader(new InputStreamReader(cliente.getInputStream()));
                 PrintWriter saida = new PrintWriter(cliente.getOutputStream(), true)) {

                System.out.println("[TCP] Cliente conectado: " + cliente.getRemoteSocketAddress());
                String mensagem;

                while ((mensagem = entrada.readLine()) != null) {
                    System.out.println("[TCP] Recebido: " + mensagem);

                    if (mensagem.equalsIgnoreCase("sair")) {
                        saida.println("Encerrando conexao. Ate mais!");
                        break;
                    }

                    saida.println(responder(mensagem));
                }
            }
        }

        System.out.println("[TCP] Servidor encerrado.");
    }

    private static String responder(String mensagem) {
        if (mensagem.equalsIgnoreCase("hora")) {
            return "Horario atual do servidor: " + LocalDateTime.now().format(FORMATO_HORA);
        }

        return "Monitor responde: recebi sua mensagem -> \"" + mensagem + "\"";
    }

    private static int lerOffset(String[] args) {
        if (args.length > 0 && args[0].matches("\\d+")) {
            return Integer.parseInt(args[0]);
        }

        String valor = System.getenv("LAB_REDES_OFFSET");
        return valor == null || valor.isBlank() ? 0 : Integer.parseInt(valor);
    }
}

File: lab-redes/python/tcp/servidor_tcp.py
from datetime import datetime
import os
import socket
import sys

HOST = "0.0.0.0"
PORTA_BASE = 5000


def ler_offset() -> int:
    if len(sys.argv) > 1 and sys.argv[1].isdigit():
        return int(sys.argv[1])
    return int(os.getenv("LAB_REDES_OFFSET", "0"))


def responder(mensagem: str) -> str:
    if mensagem.lower() == "hora":
        return f"Horario atual do servidor: {datetime.now():%H:%M:%S}"
    return f'Monitor responde: recebi sua mensagem -> "{mensagem}"'


porta = PORTA_BASE + ler_offset()

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as servidor:
    servidor.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    servidor.bind((HOST, porta))
    servidor.listen(1)
    print(f"[TCP] Servidor aguardando conexoes na porta {porta}...", flush=True)

    conexao, endereco = servidor.accept()
    with conexao:
        print(f"[TCP] Cliente conectado: {endereco}", flush=True)
        while True:
            dados = conexao.recv(1024).decode("utf-8").strip()
            if not dados:
                break

            print(f"[TCP] Recebido: {dados}", flush=True)
            if dados.lower() == "sair":
                conexao.sendall("Encerrando conexao. Ate mais!\n".encode("utf-8"))
                break

            conexao.sendall((responder(dados) + "\n").encode("utf-8"))

print("[TCP] Servidor encerrado.", flush=True)
